src/sinks/pulsar/service.rs

```diff
@@ -0,0 +1,194 @@
+            .producer()
+            .with_topic(topic)
+            .with_options(producer_options);
+        match prod.build().await {
+            Ok(p) => Ok(Arc::new(Mutex::new(p))),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Pulsar requires a producer object be created per topic
+    /// This method will build a producer if it hasn't been created or caches it otherwise
+    async fn get_or_build_producer(
+        producer_cache: SafeLru<Exe>,
+        client: Pulsar<Exe>,
+        producer_options: ProducerOptions,
+        topic: String,
+    ) -> Arc<Mutex<Producer<Exe>>> {
+        let mut pc = producer_cache.lock().await;
+        match pc.contains(&topic) {
+            false => {
+                pc.put(
+                    topic.clone(),
+                    PulsarService::build_producer(client, producer_options, &topic).await,
+                );
+                let f = pc.get(&topic).unwrap().as_ref().unwrap();
+                Arc::clone(f)
+            }
+            true => {
+                let f = pc.get(&topic).unwrap().as_ref().unwrap();
+                Arc::clone(f)
+            }
+        }
+    }
+}
+
+impl<Exe: Executor> Service<PulsarRequest> for PulsarService<Exe> {
+    type Response = PulsarResponse;
+    type Error = PulsarError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, request: PulsarRequest) -> Self::Future {
+        let prod_future = PulsarService::get_or_build_producer(
+            Arc::clone(&self.producer_cache),
+            self.pulsar_client.clone(),
+            self.producer_options.clone(),
+            request.metadata.topic.clone(),
+        );
+        let bytes_sent = self.bytes_sent.clone();
+        let ts = request.metadata.timestamp_millis.to_owned();
+        Box::pin(async move {
+            let p = prod_future.await;
+            let mut lp = p.lock().await;
+            let body = request.body.clone();
+            let mut msg_builder = lp.create_message().with_content(body.as_ref());
+            if let Some(key) = request.metadata.key {
+                msg_builder = msg_builder.with_key(String::from_utf8_lossy(&*key));
+            }
+            if let Some(timestamp) = ts {
+                msg_builder = msg_builder.event_time(timestamp as u64);
+            }
+            if let Some(properties) = request.metadata.properties {
+                for (key, value) in properties {
+                    msg_builder =
+                        msg_builder.with_property(key, String::from_utf8_lossy(&*value.clone()));
+                }
+            }
+
+            match msg_builder.send().await {
+                Ok(resp) => match resp.await {
+                    Ok(_) => {
+                        //TODO: not totally accurate, doesn't include metadata or key
```

Comment:
Just want to make sure this TODO isn't forgotten about.
Here are the requirements for what this should count: https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#componentbytessent